packages/Discretization/test/tstNearestNeighborOperator.cpp

```diff
@@ -0,0 +1,396 @@
+
+    Kokkos::realloc( target_values, target_points.extent( 0 ) );
+    Kokkos::realloc( source_values, source_points.extent( 0 ) );
+    if ( comm_rank == 0 )
+    {
+        auto source_values_host = Kokkos::create_mirror_view( source_values );
+        source_values_host( 0 ) = 255.;
+        Kokkos::deep_copy( source_values, source_values_host );
+    }
+
+    nnop.apply( source_values, target_values );
+
+    auto target_values_host = Kokkos::create_mirror_view( target_values );
+    Kokkos::deep_copy( target_values_host, target_values );
+    std::vector<double> target_values_ref = {255.};
+    TEST_COMPARE_ARRAYS( target_values_host, target_values_ref );
+
+    // Reset the static variable to its original value to avoid interfering
+    // with other unit tests.
+    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::epsilon =
+        epsilon_default;
+}
+
+TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( NearestNeighborOperator, structured_clouds,
+                                   DeviceType )
+{
+    // The source is a structured cloud. The target is the same cloud but
+    // distributed differently among the processors.
+    Teuchos::RCP<Teuchos::Comm<int> const> comm =
+        Teuchos::DefaultComm<int>::getComm();
+    unsigned int const comm_size = comm->getSize();
+    unsigned int const comm_rank = comm->getRank();
+
+    // Build the structured cloud of points for the source and the target.
+    double const L_x = 2.;
+    double const L_y = 3.;
+    double const L_z = 5.;
+    unsigned int const n_x = 7;
+    unsigned int const n_y = 11;
+    unsigned int const n_z = 13;
+    double const source_offset_x = comm_rank * L_x;
+    double const source_offset_y = comm_rank * L_y;
+    double const source_offset_z = comm_rank * L_z;
+
+    Kokkos::View<double **, DeviceType> source_points( "source" );
+    copyPointsFromCloud<DeviceType>(
+        makeStructuredCloud( L_x, L_y, L_z, n_x, n_y, n_z, source_offset_x,
+                             source_offset_y, source_offset_z ),
+        source_points );
+
+    double const target_offset_x = ( ( comm_rank + 1 ) % comm_size ) * L_x;
+    double const target_offset_y = ( ( comm_rank + 1 ) % comm_size ) * L_y;
+    double const target_offset_z = ( ( comm_rank + 1 ) % comm_size ) * L_z;
+
+    Kokkos::View<double **, DeviceType> target_points( "target" );
+    copyPointsFromCloud<DeviceType>(
+        makeStructuredCloud( L_x, L_y, L_z, n_x, n_y, n_z, target_offset_x,
+                             target_offset_y, target_offset_z ),
+        target_points );
+
+    unsigned int const n_points = source_points.extent( 0 );
+    Kokkos::View<double *, DeviceType> target_values( "target_values",
+                                                      n_points );
+
+    DataTransferKit::NearestNeighborOperator<DeviceType> nnop(
+        comm, source_points, target_points );
+
+    // Wish I could use subview but Kokkos doesn't apply's interface
```

Comment:
Missing a work `doesn't apply` -> `doesn't like apply`

Reply:
I changed the comment.  But what did you mean here?  (This was committed by you ;))

Reply:
Ideally, I wouldn't need the `parallel_for` to build `source_values`. `source_values` would be a subview of `source_points` but it doesn't compile because of `apply`.

Reply:
Should we template `apply` over the View types then?